File: platform/sponge/src/main/java/ca/stellardrift/permissionsex/sponge/PEXSubjectData.java
/*
 * PermissionsEx
 * Copyright (C) zml and PermissionsEx contributors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package ca.stellardrift.permissionsex.sponge;

import ca.stellardrift.permissionsex.context.ContextValue;
import ca.stellardrift.permissionsex.impl.util.PCollections;
import ca.stellardrift.permissionsex.subject.Segment;
import ca.stellardrift.permissionsex.subject.SubjectRef;
import ca.stellardrift.permissionsex.util.Change;
import com.google.common.collect.Maps;
import org.checkerframework.checker.nullness.qual.Nullable;
import org.spongepowered.api.service.context.Context;
import org.spongepowered.api.service.permission.Subject;
import org.spongepowered.api.service.permission.SubjectData;
import org.spongepowered.api.service.permission.SubjectReference;
import org.spongepowered.api.util.Tristate;

import java.util.List;
import java.util.Map;
import java.util.Set;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.ConcurrentMap;

import static java.util.Objects.requireNonNull;

public class PEXSubjectData implements SubjectData {
    private final SubjectRef.ToData<?> data;
    private final PEXSubject subject;
    private final ConcurrentMap<Set<ContextValue<?>>, List<SubjectReference>> parentsCache = new ConcurrentHashMap<>();


    PEXSubjectData(final SubjectRef.ToData<?> data, final PEXSubject subject) {
        this.data = data;
        this.subject = subject;
        data.onUpdate($ -> this.clearCache());
    }

    PermissionsExService service() {
        return this.subject.getContainingCollection().service();
    }

    private void clearCache() {
        synchronized(this.parentsCache) {
            this.parentsCache.clear();
        }
    }

    private <T> Map<Set<Context>, T> keysToSponge(final Map<Set<ContextValue<?>>, T> original) {
        return PCollections.asMap(original, (k, $) -> Contexts.toSponge(k), ($, v) -> v);
    }

    @Override
    public Subject getSubject() {
        return this.subject;
    }

    @Override
    public boolean isTransient() {
        return this == this.subject.getTransientSubjectData();
    }

    @Override
    public Map<Set<Context>, Map<String, Boolean>> getAllPermissions() {
        return keysToSponge(data.get().mapSegmentValues(it -> Maps.transformValues(it.permissions(), v -> v > 0)));
    }

    @Override
    public Map<String, Boolean> getPermissions(final Set<Context> contexts) {
        return Maps.transformValues(this.data.get().segment(Contexts.toPex(contexts, service().manager()))
            .permissions(), v -> v > 0);
    }

    @Override
    public CompletableFuture<Boolean> setPermission(
        final Set<Context> contexts,
        final String permission,
        final Tristate value
    ) {
        final int intVal;
        switch (value) {
            case TRUE: intVal = 1; break;
            case FALSE: intVal = -1; break;
            case UNDEFINED: intVal = 0; break;
            default: throw new IllegalStateException("Unknown tristate provided: " + value);
        }
        return data.update(Contexts.toPex(contexts, service().manager()), it -> it.withPermission(permission, intVal))
            .thenApply( $ -> true);
    }

    @Override
    public CompletableFuture<Boolean> clearPermissions() {
        return data.update(it -> it.withSegments(($, segment) -> segment.withoutPermissions()))
            .thenApply(Change::changed);
    }

    @Override
    public CompletableFuture<Boolean> clearPermissions(final Set<Context> contexts) {
        return data.update(Contexts.toPex(contexts, service().manager()), Segment::withoutPermissions)
            .thenApply(Change::changed);
    }

    @Override
    public Map<Set<Context>, List<SubjectReference>> getAllParents() {
        synchronized(parentsCache) {
            this.data.get().activeContexts().forEach(this::getParentsInternal);
            return keysToSponge(this.parentsCache);
        }
    }

    @Override
    public List<SubjectReference> getParents(final Set<Context> contexts) {
        return this.getParentsInternal(Contexts.toPex(contexts, service().manager()));
    }

    private List<SubjectReference> getParentsInternal(final Set<ContextValue<?>> contexts) {
        final List<SubjectReference> existing = this.parentsCache.get(contexts);
        if (existing != null) {
            return existing;
        }
        final List<SubjectReference> parents;
        synchronized(this.parentsCache) {
            final List<? extends SubjectRef<?>> rawParents = this.data.get().segment(contexts).parents();
            parents = rawParents == null ? PCollections.vector() : PCollections.asVector(rawParents, it -> PEXSubjectReference.asSponge(it, this.service()));
            final @Nullable List<SubjectReference> existingParents = this.parentsCache.putIfAbsent(contexts, parents);
            if (existingParents != null) {
                return existingParents;
            }
        }
        return parents;
    }

    @Override
    public CompletableFuture<Boolean> addParent(final Set<Context> contexts, final SubjectReference parent) {
        requireNonNull(contexts, "contexts");
        requireNonNull(parent, "parent");
        final PEXSubjectReference<?> ref = PEXSubjectReference.asPex(parent, this.service()); // validate subject reference
        return data.update(Contexts.toPex(contexts, service().manager()), it -> it.plusParent(ref))
            .thenApply(Change::changed);
    }

    @Override
    public CompletableFuture<Boolean> removeParent(final Set<Context> contexts, final SubjectReference parent) {
        requireNonNull(parent, "parent");
        return data.update(Contexts.toPex(contexts, service().manager()), it -> it.minusParent(PEXSubjectReference.asPex(parent, this.service())))
            .thenApply(Change::changed);
    }

    @Override
    public CompletableFuture<Boolean> clearParents() {
        return data.update(it -> it.withSegments(($, segment) -> segment.withoutParents()))
            .thenApply(Change::changed);
    }

    @Override
    public CompletableFuture<Boolean> clearParents(final Set<Context> contexts) {
        return data.update(Contexts.toPex(contexts, service().manager()), Segment::withoutParents)
            .thenApply(Change::changed);
    }

    @Override
    public Map<Set<Context>, Map<String, String>> getAllOptions() {
        return keysToSponge(this.data.get().mapSegmentValues(Segment::options));
    }

    @Override
    public Map<String, String> getOptions(final Set<Context> contexts) {
        return data.get().segment(Contexts.toPex(contexts, service().manager())).options();
    }

    @Override
    public CompletableFuture<Boolean> setOption(
        final Set<Context> contexts,
        final String key,
        final @Nullable String value
    ) {
        return data.update(Contexts.toPex(contexts, service().manager()),
            it -> value == null ? it.withoutOption(key) : it.withOption(key, value))
            .thenApply(Change::changed);
    }

    @Override
    public CompletableFuture<Boolean> clearOptions() {
        return data.update(it -> it.withSegments(($, segment) -> segment.withoutOptions()))
            .thenApply(Change::changed);
    }

    @Override
    public CompletableFuture<Boolean> clearOptions(final Set<Context> contexts) {
        return data.update(Contexts.toPex(contexts, service().manager()), Segment::withoutOptions)
            .thenApply(Change::changed);
    }

}

File: platform/sponge/src/main/kotlin/ca/stellardrift/permissionsex/sponge/PEXSubjectData.kt
/*
 * PermissionsEx
 * Copyright (C) zml and PermissionsEx contributors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package ca.stellardrift.permissionsex.sponge

import ca.stellardrift.permissionsex.data.Change
import ca.stellardrift.permissionsex.data.ImmutableSubjectData
import ca.stellardrift.permissionsex.data.SubjectDataReference
import ca.stellardrift.permissionsex.util.ContextSet
import java.util.concurrent.CompletableFuture
import java.util.concurrent.ConcurrentHashMap
import java.util.concurrent.ConcurrentMap
import org.spongepowered.api.service.context.Context
import org.spongepowered.api.service.permission.Subject
import org.spongepowered.api.service.permission.SubjectData
import org.spongepowered.api.service.permission.SubjectReference
import org.spongepowered.api.util.Tristate

/**
 * Wrapper around ImmutableSubjectData that writes to backend each change
 */
class PEXSubjectData internal constructor(
    private val data: SubjectDataReference,
    private val subject: PEXSubject
) : SubjectData {
    private val parentsCache: ConcurrentMap<ContextSet, List<SubjectReference>> = ConcurrentHashMap()
    private val service: PermissionsExService = subject.collection.service

    init {
        data.onUpdate { clearCache() }
    }

    /**
     * Provide a boolean representation of success for the Sponge returns.
     *
     * @return Whether or not the old data object is different from the new data object
     */
    private fun CompletableFuture<Change<ImmutableSubjectData?>>.boolSuccess(): CompletableFuture<Boolean> {
        return thenApply { it.old != it.new }
    }

    private fun clearCache() {
        synchronized(parentsCache) { parentsCache.clear() }
    }

    override fun getAllOptions(): Map<Set<Context>, Map<String, String>> {
        return data.get().allOptions.keysToSponge()
    }

    override fun getOptions(contexts: Set<Context>): Map<String, String> {
        return data.get().getOptions(contexts.toPex(service.manager))
    }

    override fun setOption(contexts: Set<Context>, key: String, value: String?): CompletableFuture<Boolean> {
        return data.update { it.setOption(contexts.toPex(service.manager), key, value) }.boolSuccess()
    }

    override fun clearOptions(contexts: Set<Context>): CompletableFuture<Boolean> {
        return data.update { it.clearOptions(contexts.toPex(service.manager)) }.boolSuccess()
    }

    override fun clearOptions(): CompletableFuture<Boolean> {
        return data.update { obj: ImmutableSubjectData -> obj.clearOptions() }.boolSuccess()
    }

    override fun getSubject(): Subject {
        return this.subject
    }

    override fun isTransient(): Boolean {
        return this === this.subject.transientSubjectData
    }

    override fun getAllPermissions(): Map<Set<Context>, Map<String, Boolean>> {
        return data.get().allPermissions
            .keysToSponge()
            .mapValues { (_, v) -> v.mapValues { it.value > 0 } }
    }

    override fun getPermissions(contexts: Set<Context>): Map<String, Boolean> {
        return data.get().getPermissions(contexts.toPex(service.manager))
            .mapValues { (_, v) -> v > 0 }
    }

    override fun setPermission(contexts: Set<Context>, permission: String, state: Tristate): CompletableFuture<Boolean> {
        val value = when (state) {
            Tristate.TRUE -> 1
            Tristate.FALSE -> -1
            Tristate.UNDEFINED -> 0
            else -> throw IllegalStateException("Unknown tristate provided $state")
        }
        return data.update { it.setPermission(contexts.toPex(service.manager), permission, value) }.thenApply { true }
    }

    override fun clearPermissions(): CompletableFuture<Boolean> {
        return data.update { it.clearPermissions() }.boolSuccess()
    }

    override fun clearPermissions(contexts: Set<Context>): CompletableFuture<Boolean> {
        return data.update { it.clearPermissions(contexts.toPex(service.manager)) }.boolSuccess()
    }

    override fun getAllParents(): Map<Set<Context>, List<SubjectReference>> {
        synchronized(parentsCache) {
            data.get().activeContexts.forEach { getParentsInternal(it) }
            return parentsCache.keysToSponge()
        }
    }

    override fun getParents(contexts: Set<Context>): List<SubjectReference> {
        return getParentsInternal(contexts.toPex(service.manager))
    }

    private fun getParentsInternal(contexts: ContextSet): List<SubjectReference> {
        val existing = parentsCache[contexts]
        if (existing != null) {
            return existing
        }
        val parents: List<SubjectReference>
        synchronized(parentsCache) {
            val rawParents = data.get().getParents(contexts)
            parents = if (rawParents == null) {
                emptyList()
            } else {
                rawParents.map { (k, v) -> service.newSubjectReference(k, v) }
            }
            val existingParents = parentsCache.putIfAbsent(contexts, parents)
            if (existingParents != null) {
                return existingParents
            }
        }
        return parents
    }

    override fun addParent(contexts: Set<Context>, subject: SubjectReference): CompletableFuture<Boolean> {
        subject.asPex(service) // validate subject reference
        return data.update {
            it.addParent(contexts.toPex(service.manager), subject.collectionIdentifier, subject.subjectIdentifier)
        }.boolSuccess()
    }

    override fun removeParent(set: Set<Context>, subject: SubjectReference): CompletableFuture<Boolean> {
        return data.update {
            it.removeParent(set.toPex(service.manager), subject.collectionIdentifier, subject.subjectIdentifier)
        }.boolSuccess()
    }

    override fun clearParents(): CompletableFuture<Boolean> {
        return data.update { it.clearParents() }.boolSuccess()
    }

    override fun clearParents(set: Set<Context>): CompletableFuture<Boolean> {
        return data.update { it.clearParents(set.toPex(service.manager)) }.boolSuccess()
    }
}

private fun <T> Map<ContextSet, T>.keysToSponge(): Map<Set<Context>, T> {
    return mapKeys { (key, _) -> key.toSponge() }
}
